### DISCUSSION TITLE
Relationships between Tenants

### DISCUSSION THREAD

**Discussion by stephen776** [JonPSmith/AuthPermissions.AspNetCore#48] [Q&A]
Hello,

First, awesome library. I love the ideas proposed here and I’ve run into some of these problems in the past.

I’m trying to figure out how I might design a system that consists of many tenants (each with many users) where tenants have explicit relationships defined (along the lines of a “friends” list relationship)

the relationship between tenants would enable Tenant A to send a “work request” to their “friend” Tenant B

tenant B would be able to view the request and assign their users to work on it before updating the record at which time Tenant A could review it.

i was considering something along the lines of either:

creating temporary sub-tenants where Tenant B could become a sub tenant of A during the time in which there was a “request” from A that needed attention 

or

I’m thinking is there potentially a way for a user to “join” a different “friend” tenant temporarily to perform some work before “returning” to their “home tenant”

maybe there’s some other approach I haven’t thought of yet.

curious to get your input

**Comment by JonPSmith** [comment]
Hi @stephen776,

I understand what you are trying to create - its a bit like GitHub, WhatsApp, etc. I haven't though about adding this type of multi-tenant because the design of these types all have different features.

I'm sure you could come up with a design that works for your specific requirement but the issue will be performance, so solve that part first. Of the top of my head you could have a collection of `DataKeys` (see first paragraph of the [Multi-tenant-explained](https://github.com/JonPSmith/AuthPermissions.AspNetCore/wiki/Multi-tenant-explained ) docs for `DataKey`), instead of a single `DataKey`. Then you could create a query filter where the query is something like this:

```c#
modelBuilder.Entity<Invoice>().HasQueryFilter(x => x.DataKeys.Contains(userDataKey));
```

All the best with your project. If you blog about tech things, then please write about your solution and send me a link. I would be interested.
